README.md
# vy_date

Simple utilities for managing dates and periods.

## Date class

The Date class simply wraps the DateTime one.
I believe that it is easier to manage dates this way because it frees you from the hassle of the time zones and prevents you doing the following potentially dangerous moves:

- Involuntarily set minutes or seconds or so on.
This can lead to problems when comparing dates
- Have to work with dates from different time zones or mixed with UTC ones.

Pluses:

- You can be sure that the difference in days is always correct.
- Have some easy to use functions for formatting
- Json formatting

Instances can be serialized with Json, in this case they are sent as a `String`
with the date in format "YYYY-MM-DD".

Technically speaking, the date is transformed int an UTC DateTime instance.
During insertion all references to time info are lost.

There are a couple of formatting functions and their counterparts for parsing the result. These methods **require** `Intl` to be initialized.
Inside your application you have to import the `Intl` package:

`import 'package:intl/date_symbol_data_local.dart';`

and you have to initialize the locale you'll gonna use with the statement:

`await initializeDateFormatting('en_US', null);`

This is needed only once in the whole application, before using the formatting methods for the rquired locale, otherwise a `LocaleDataException` will be thrown

### Usage

A simple usage example:

    import 'package:vy_date/vy_date.dart';

    main() {
      Date date = Date.now();
    }

## DatePeriod class

This class allows to manage a period between two dates.
It is immutable, and has a companion class that allows to manage partial
data before generating a correct instance of DatePeriod.
This companion class is called DatePeriodAssembler and follows the logic of the Builder classes. I could have used the name DatePeriodBuilder,
but I did not want to confuse how the class behave with those of the `built_value` package. The logic is + or - the same, but there is a difference: the data class (DatePeriod in this case) is not aware of the existence of an helper method, and could even be used alone, while the built_value
package generated classes are tightly coupled together.

The class is comparable, the comparision is made first testing the start date
and then, if equal, testing the period duration.

The DatePeriod class does not accept negative periods (i.e. those with the end
period date before in time of the start period one).

You can even initialize the period with null values (even if this is discouraged).
If a null start date is receive the conventional date 1970-01-01 is used,
if the end date (or duration, depending on the constructor) is null,
a period of one day is used.

The class is normally managed with initial and ending dates inclusive, but you can extract the exclusive end date with the appropriate getter if needed.

When converted to Json string the class is saved as Map containing the two dates into the above (see Date class) mentioned format.

### DatePeriod Usage

A simple usage example:

    import 'package:vy_date/vy_date.dart';

    void main() {
      DatePeriod period = DatePeriod(Date(2010, 1, 1), Date(2010, 1, 31));
      // print "31"
      print(period.inDays);
      
      DatePeriodAssembler periodAssembler = DatePeriodAssembler()
        ..startDate = Date(2010, 1, 1)
        ..exclusiveEndDate = Date(2010, 2, 1);
      // print "0"
      print(periodAssembler.generate().compareTo(period)); 
    }

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/giorgiofran/vy_date/issues

[License](https://github.com/giorgiofran/vy_date/blob/master/LICENSE).


CHANGELOG.md
# Changelog

## 1.0.2

- updated dependencies
- removed quiver dependency

## 1.0.1

- Lints

## 1.0.0-nullsafety

Pre-release for non nullable data types.

### Breaking changes

- Date.parse() now throws an error if the parsed string is empty instead of returning null. A tryParse method is provided to maintain the old functionality.

## 0.4.3

- License update

## 0.4.2

- Lint cleanup

## 0.4.1

- Package Intl ^0.16.0 instead of ^0.15.7
- Added toYMMddString() and parseYMMddString().
  The behavior of these functions is to format and parse a date in a way similar to YMd(), with the difference
  that month and day are padded with 0 when the length is 1.
  In some countries this is a legal requirement for official documents.
  
## 0.4.0

- DatePeriod transformed into an immutable class
- Removed deprecated setter and getter for "startingDate"
- Removed deprecated setter and getter for "endingDate"
- Removed deprecated "getEndDate" method
- Removed deprecated "getFirstExcludedDay" method
- Removed deprecated "getDays" method
- Static constructors have been transformed into factories

## 0.3.4

- Added toYMdString() and parseYMdString()

## 0.3.3

- Sdk 2.2.0 support
- Call to DatePeriod.isInPeriod() was giving error if the date parameter was null. Now returns false.
- Changed the way how the DatePeriod is serialized (toJson). Instead of saving the duration, it is saved the end date in format yyyy-mm-dd
- Added the weekday getter

## 0.3.2

- Sdk 2.1.0 support.
- Some Linter rules applied  
  
## 0.3.1

- Dart 2 support

## 0.3.0

- Initial version


test/date_test.dart
import 'package:intl/date_symbol_data_local.dart';
import 'package:vy_date/vy_date.dart' show Date;
import 'package:test/test.dart';

void main() async {
  await initializeDateFormatting('it_IT', null);

  group('Date Creation', () {
    test('Standard constructor', () {
      final dt = Date(2008, 12, 29);
      final dt2 = Date(-15, 01, 29);
      expect(dt.day, 29);
      expect(dt2.year, -15);
      expect(dt2.month, 01);
    });
    test('now', () {
      final date = Date.now();
      final dt = DateTime.now();
      final dt2 = Date(2015, 02, 03);
      expect(date.year, dt.year);
      expect(date.month, dt.month);
      expect(date.day, dt.day);
      expect(Date.now().isAfter(dt2), isTrue);
    });
    test('fromDateTime', () {
      final date = Date.now();
      final date2 = Date.fromDateTime(DateTime.now());
      expect(date, date2);
    });
    test('static - parse', () {
      var date = Date.parse('2012-02-27');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('2012-02-27 13:27:00');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('20120227 13:27:00');
      expect(date, Date(2012, 02, 27));

      date = Date.parse('20120227T132700');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('20120227');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('+20120227');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('2012-02-27T14Z');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('2012-02-27T14+00:00');
      expect(date, Date(2012, 02, 27));
      date = Date.parse('-123450101 00:00:00 Z');
      expect(date, Date(-12345, 01, 01));
      date = Date.parse('2002-02-27T14:00:00-0500');
      expect(date, Date(2002, 02, 27));
    });
  });

  group('Output', () {
    test('Iso 8601', () {
      final dt = Date(2008, 12, 29);
      final dt2 = Date(-15, 01, 29);
      expect(dt.toIso8601String(), '2008-12-29');
      expect(dt2.toIso8601String(), '-0015-01-29');
    });
    test('To String', () {
      final dt = Date(2008, 12, 29);
      final dt2 = Date(-15, 01, 29);
      expect(dt.toString(), '2008-12-29');
      expect(dt2.toString(), '-0015-01-29');
    });
  });

  group('Comparison', () {
    test('equals', () {
      final dt = Date(2018, 02, 03);
      expect(Date(2018, 02, 03), dt);
    });

    test('CompareTo', () {
      final dt = Date(2008, 12, 29);
      final dt3 = Date.now();
      final dt2 = Date(-15, 01, 29);
      expect(dt3.compareTo(dt), 1);
      expect(dt2.compareTo(dt), -1);
      expect(dt.compareTo(dt), 0);
    });
    test('is Before', () {
      final dt = Date(2008, 12, 29);
      final dt2 = Date(-15, 01, 29);
      final dt3 = Date.now();

      expect(dt.isBefore(dt3), isTrue);
      expect(dt2.isBefore(dt), isTrue);
      expect(dt3.isBefore(dt2), isFalse);
    });
    test('is After', () {
      final dt = Date(2008, 12, 29);
      final dt2 = Date(-15, 01, 29);
      final dt3 = Date.now();

      expect(dt.isAfter(dt3), isFalse);
      expect(dt.isAfter(dt2), isTrue);
      expect(dt3.isAfter(dt2), isTrue);
    });

    test('toJson', () {
      final dt = Date(2018, 02, 03);
      expect(dt.toJson(), '2018-02-03');
    });
  });

  group('Formatting', () {
    test('Format to String', () {
      final dt = Date(2018, 02, 03);
      expect(dt.toYMMMMdString('en_US'), 'February 3, 2018');
      expect(dt.toYMMMMdString('it_IT'), '3 febbraio 2018');
      expect(dt.toYMMMMdString('te_IN'), '3 ఫిబ్రవరి, 2018');
    });

    test('Parse from String', () {
      final dt = Date(2018, 02, 03);
      expect(Date.parseYMMMMdString('3 febbraio 2018', 'it_IT'), dt);
      expect(Date.parseYMMMMdString('February 3, 2018', 'en_US'), dt);
      expect(Date.parseYMMMMdString('3 ఫిబ్రవరి, 2018', 'te_IN'), dt);
    });

    test('Format to String', () {
      final dt = Date(2018, 02, 03);
      expect(dt.toYMdString('en_US'), '2/3/2018');
      expect(dt.toYMdString('it_IT'), '3/2/2018');
      expect(dt.toYMdString('te_IN'), '3/2/2018');
    });

    test('Parse from String', () {
      final dt = Date(2018, 02, 03);
      expect(Date.parseYMdString('3/2/2018', 'it_IT'), dt);
      expect(Date.parseYMdString('2/3/2018', 'en_US'), dt);
      expect(Date.parseYMdString('3/2/2018', 'te_IN'), dt);
    });

    test('Format to String yMMdd', () {
      final dt = Date(2018, 02, 03);
      expect(dt.toYMMddString('en_US'), '02/03/2018');
      expect(dt.toYMMddString('it_IT'), '03/02/2018');
      expect(dt.toYMMddString('te_IN'), '03/02/2018');
    });

    test('Parse from String yMMdd', () {
      final dt = Date(2018, 02, 03);
      expect(Date.parseYMMddString('03/02/2018', 'it_IT'), dt);
      expect(Date.parseYMMddString('02/03/2018', 'en_US'), dt);
      expect(Date.parseYMMddString('03/02/2018', 'te_IN'), dt);
    });
  });
}


test/date_period_test.dart
import 'package:intl/date_symbol_data_local.dart';
import 'package:vy_date/vy_date.dart'
    show Date, DatePeriod, DatePeriodAssembler;
import 'package:test/test.dart';

void main() async {
  await initializeDateFormatting('it_IT', null);
  DatePeriod datePeriod, datePeriod_2;
  DatePeriodAssembler datePeriodAssembler, datePeriodAssembler_2;

  final date = Date(2008, 12, 29);

  group('DatePeriod Creation', () {
    test('Standard constructor', () {
      datePeriodAssembler = DatePeriodAssembler();
      expect(datePeriodAssembler.startDate, isNull);
      expect(datePeriodAssembler.duration, isNull);

      datePeriodAssembler = DatePeriodAssembler(date);
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.duration, isNull);
      expect(datePeriodAssembler.isValid(), isFalse);

      datePeriodAssembler = DatePeriodAssembler(date, Date(2009, 1, 7));
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.duration, Duration(days: 10));
      expect(datePeriodAssembler.endDate, Date(2009, 1, 7));
      expect(datePeriodAssembler.isValid(), isTrue);
      expect(datePeriodAssembler.exclusiveEndDate, Date(2009, 1, 8));

      datePeriod = DatePeriod(date, Date(2009, 1, 7));
      expect(datePeriod.startDate, date);
      expect(datePeriod.duration, Duration(days: 10));
      expect(datePeriod.endDate, Date(2009, 1, 7));
      expect(datePeriod.exclusiveEndDate, Date(2009, 1, 8));

      datePeriod = DatePeriodAssembler(date, Date(2009, 1, 7)).generate();
      expect(datePeriod.startDate, date);
      expect(datePeriod.duration, Duration(days: 10));
      expect(datePeriod.endDate, Date(2009, 1, 7));
      expect(datePeriod.exclusiveEndDate, Date(2009, 1, 8));
    });

    test('ByDuration constructor', () {
      datePeriod = DatePeriod.byDuration(date, Duration(days: 10));
      expect(datePeriod.startDate, date);
      expect(datePeriod.duration, Duration(days: 10));
      expect(datePeriod.exclusiveEndDate, Date(2009, 1, 8));
      expect(datePeriod.endDate, Date(2009, 1, 7));
    });
  });

  group('Setting values', () {
    test('Setting startDate', () {
      datePeriodAssembler = DatePeriodAssembler()..startDate = date;
      expect(datePeriodAssembler.duration, isNull);
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.isValid(), isFalse);

      datePeriodAssembler.duration = Duration(days: 10);
      expect(datePeriodAssembler.exclusiveEndDate, Date(2009, 1, 8));

      expect(() {
        datePeriodAssembler.startDate = Date(2009, 1, 10);
      }, throwsArgumentError);

      datePeriodAssembler.startDate = Date(2009, 1, 1);
      expect(datePeriodAssembler.duration, Duration(days: 7));
      expect(datePeriodAssembler.endDate, Date(2009, 1, 7));
    });

    test('Setting duration', () {
      datePeriodAssembler = DatePeriodAssembler()
        ..duration = Duration(days: 10);
      expect(datePeriodAssembler.startDate, isNull);
      expect(datePeriodAssembler.duration, Duration(days: 10));
      expect(datePeriodAssembler.isValid(), isFalse);

      datePeriodAssembler.startDate = date;
      expect(datePeriodAssembler.exclusiveEndDate, Date(1970, 1, 11));

      expect(() {
        datePeriodAssembler.duration = Duration(days: 0);
      }, throwsArgumentError);
      expect(() {
        datePeriodAssembler.duration = Duration(days: -1);
      }, throwsArgumentError);

      datePeriodAssembler.duration = Duration(days: 3, hours: 23);
      expect(datePeriodAssembler.duration, Duration(days: 3));
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.endDate, Date(2008, 12, 31));
    });

    test('Setting exclusive end date', () {
      datePeriodAssembler = DatePeriodAssembler()
        ..exclusiveEndDate = Date(2009, 1, 8);
      expect(datePeriodAssembler.startDate, null);
      expect(datePeriodAssembler.duration, Duration(hours: 342048));
      expect(datePeriodAssembler.isValid(), isFalse);

      datePeriodAssembler.startDate = date;
      expect(datePeriodAssembler.exclusiveEndDate, Date(2009, 1, 8));
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.duration, Duration(days: 10));

      expect(() {
        datePeriodAssembler.exclusiveEndDate = Date(2008, 1, 1);
      }, throwsArgumentError);

      datePeriodAssembler.duration = Duration(days: 3);
      expect(datePeriodAssembler.duration, Duration(days: 3));
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.endDate, Date(2008, 12, 31));
    });

    test('Setting end date', () {
      datePeriodAssembler = DatePeriodAssembler()..endDate = Date(2009, 1, 7);
      expect(datePeriodAssembler.startDate, null);
      expect(datePeriodAssembler.duration, Duration(hours: 342048));
      expect(datePeriodAssembler.isValid(), isFalse);

      datePeriodAssembler.startDate = date;
      expect(datePeriodAssembler.endDate, Date(2009, 1, 7));
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.duration, Duration(days: 10));

      expect(() {
        datePeriodAssembler.endDate = Date(2008, 1, 1);
      }, throwsArgumentError);

      datePeriodAssembler.duration = Duration(days: 3);
      expect(datePeriodAssembler.duration, Duration(days: 3));
      expect(datePeriodAssembler.startDate, date);
      expect(datePeriodAssembler.exclusiveEndDate, Date(2009, 1, 1));
    });
  });

  group('Checking methods', () {
    final datePeriod = DatePeriod(date, Date(2009, 1, 7));
    datePeriodAssembler = DatePeriodAssembler()..initFrom(datePeriod);
    datePeriodAssembler_2 = DatePeriodAssembler(date, null);

    final jsonMap = <String, String>{
      DatePeriod.fieldStartDate: '2008-12-29',
      DatePeriod.fieldEndDate: '2009-01-07'
    };
    final jsonMap2 = <String, dynamic>{
      DatePeriod.fieldStartDate: '2008-12-29',
      DatePeriod.fieldEndDate: '2008-12-31'
    };
    const encodedString = '{"startDate":"2008-12-29","endDate":"2009-01-07"}';
    const encodedString_2 = '{"startDate":"2008-12-29","endDate":"2008-12-31"}';

    test('to String', () {
      expect('$datePeriod', 'From 2008-12-29 to 2009-01-07 ');
      expect('$datePeriodAssembler_2', 'From 2008-12-29 to null ');
    });

    test('toJson', () {
      expect(datePeriod.toJson(), jsonMap);
    });

    test('initFrom', () {
      var dp = DatePeriodAssembler();
      dp.initFrom(datePeriod);
      expect(dp.generate(), datePeriod);
      dp = DatePeriodAssembler();
      dp.initFrom((datePeriodAssembler_2.duplicate()
            ..endDate = Date(2009, 01, 01))
          .generate());
      expect(dp != datePeriodAssembler_2, isTrue);
    });

    test('newFromMap', () {
      var dp = DatePeriodAssembler.newFromMap(jsonMap);
      expect(dp.generate(), datePeriod);
      dp = DatePeriodAssembler.newFromMap(jsonMap2);
      expect(dp,
          (datePeriodAssembler_2.duplicate()..endDate = Date(2008, 12, 31)));
    });

    test('duplicate', () {
      expect(datePeriod.duplicate(), datePeriod);
      var dp = DatePeriodAssembler()..initFrom(datePeriod.duplicate());
      expect(dp.generate(), datePeriod);
      dp = datePeriodAssembler.duplicate();
      expect(dp, datePeriodAssembler);
      dp = datePeriodAssembler_2.duplicate();
      expect(dp, datePeriodAssembler_2);
    });

    test('encode', () {
      expect(datePeriod.encode(), encodedString);
    });

    test('revive', () {
      var dp = DatePeriod.revive(encodedString);
      expect(dp, datePeriod);
      datePeriod_2 = DatePeriod(Date(2008, 12, 29), Date(2008, 12, 31));
      dp = DatePeriod.revive(encodedString_2);
      expect(dp, datePeriod_2);
    });

    test('isValid', () {
      expect(datePeriodAssembler_2.isValid(), isFalse);
    });

    test('inDays', () {
      expect(datePeriod.inDays, 10);
    });

    test('isInPeriod', () {
      expect(datePeriod.isInPeriod(date), isTrue);
      expect(datePeriod.isInPeriod(datePeriod.endDate), isTrue);
      expect(datePeriod.isInPeriod(datePeriod.exclusiveEndDate), isFalse);
      expect(datePeriod.isInPeriod(Date(2009, 1, 1)), isTrue);
      expect(datePeriod.isInPeriod(Date(2007, 12, 31)), isFalse);
      expect(datePeriod.isInPeriod(Date(2009, 12, 31)), isFalse);
    });

    test('CompareTo', () {
      var dp = DatePeriodAssembler()..initFrom(datePeriod);
      expect(datePeriod.compareTo(dp.generate()), 0);
      dp.exclusiveEndDate = Date(2009, 1, 1);
      expect(datePeriod.compareTo(dp.generate()), 1);
      dp.exclusiveEndDate = Date(2009, 1, 31);
      expect(datePeriod.compareTo(dp.generate()), -1);
      dp.startDate = Date(2008, 1, 31);
      expect(datePeriod.compareTo(dp.generate()), 1);
      dp.startDate = Date(2008, 12, 31);
      expect(datePeriod.compareTo(dp.generate()), -1);
      dp = datePeriodAssembler_2.duplicate();
      //expect(datePeriod_2.compareTo(dp.generate()), 0);
      expect(() => dp.generate(), throwsArgumentError);
      dp.endDate = Date(2009);
      expect(datePeriod.compareTo(dp.generate()), 1);
    });

    test('Comparison Operators', () {
      var dp = DatePeriodAssembler()..initFrom(datePeriod);
      expect(datePeriod >= dp.generate(), isTrue);
      dp.exclusiveEndDate = Date(2009, 1, 1);
      expect(datePeriod > dp.generate(), isTrue);
      dp.exclusiveEndDate = Date(2009, 1, 31);
      expect(datePeriod < dp.generate(), isTrue);
      dp.startDate = Date(2008, 1, 31);
      expect(datePeriod < dp.generate(), isFalse);
      dp.startDate = Date(2008, 12, 31);
      expect(datePeriod >= dp.generate(), isFalse);
      dp = datePeriodAssembler_2.duplicate();
      //expect(datePeriod_2 >= dp.generate(), isTrue);
      expect(() => dp.generate(), throwsArgumentError);
      dp.endDate = Date(2009);
      expect(datePeriod >= dp.generate(), isTrue);
    });

    test('Clearing start date', () {
      final dp = DatePeriodAssembler(Date(2008, 1, 1), Date(2009, 1, 7));
      dp.startDate = null;
      expect(dp.endDate, Date(2009, 1, 7));
      expect(dp.duration, Duration(hours: 342048));
      dp.startDate = Date(2009, 1, 1);
      expect(dp.endDate, Date(2009, 1, 7));
      Duration saveDuration;
      saveDuration = dp.duration!;
      dp.endDate = null;
      dp.endDate = Date(2009, 1, 7);
      expect(dp.duration, saveDuration);
    });

    test('Split', () {
      final dp = DatePeriod(Date(2015, 07, 09), Date(2018, 3, 14));
      List<DatePeriod> periods;
      periods = dp.splitByEndDate(Date(2010, 06, 30));
      expect(periods.length, 3);
      expect(periods[0].startDate, dp.startDate);
      expect(periods[1].startDate, Date(2016, 7, 1));
      expect(periods[2].startDate, Date(2017, 7, 1));
      expect(periods[0].endDate, Date(2016, 6, 30));
      expect(periods[1].endDate, Date(2017, 6, 30));
      expect(periods[2].endDate, dp.endDate);
    });

    test('Split', () {
      final dp = DatePeriod(Date(2015, 07, 09), Date(2018, 07, 09));
      List<DatePeriod> periods;
      periods = dp.splitByEndDate(Date(2010, 06, 30));
      expect(periods.last, DatePeriod(Date(2018, 07), Date(2018, 07, 09)));
    });

    test('Split single', () {
      final dp = DatePeriod(Date(2015, 07, 09), Date(2016, 3, 14));
      List<DatePeriod> periods;
      periods = dp.splitByEndDate(Date(2010, 06, 30));
      expect(periods.length, 1);
      expect(periods[0].startDate, dp.startDate);
      expect(periods[0].endDate, dp.endDate);
    });

    test('Split same enda date', () {
      final dp = DatePeriod(Date(2015, 07, 09), Date(2017, 06, 30));
      List<DatePeriod> periods;
      periods = dp.splitByEndDate(Date(2010, 06, 30));
      expect(periods.length, 2);
      expect(periods[0].startDate, dp.startDate);
      expect(periods[1].startDate, Date(2016, 07, 1));
      expect(periods[0].endDate, Date(2016, 06, 30));
      expect(periods[1].endDate, dp.endDate);
    });
  });

  group('Formatting', () {
    final datePeriod = DatePeriod(date, Date(2009, 1, 7));

    test('Format to String', () {
      expect(datePeriod.toYMMMMdString('en_US'),
          'December 29, 2008 ➖ January 7, 2009');
      expect(datePeriod.toYMMMMdString('it_IT'),
          '29 dicembre 2008 ➖ 7 gennaio 2009');
      expect(datePeriod.toYMMMMdString('te_IN'),
          '29 డిసెంబర్, 2008 ➖ 7 జనవరి, 2009');
    });

    test('Parse from String', () {
      expect(
          DatePeriod.parseYMMMMdString(
              '29 dicembre 2008 ➖ 7 gennaio 2009', 'it_IT'),
          datePeriod);
      expect(
          DatePeriod.parseYMMMMdString(
              'December 29, 2008 ➖ January 7, 2009', 'en_US'),
          datePeriod);
      expect(
          DatePeriod.parseYMMMMdString(
              '29 డిసెంబర్, 2008 ➖ 7 జనవరి, 2009', 'te_IN'),
          datePeriod);
    });
  });
}


example/example.dart
import 'package:intl/date_symbol_data_local.dart' show initializeDateFormatting;

import 'package:vy_date/vy_date.dart';

void main() async {
  var date = Date.now();
  final nextDay = date.add(Duration(days: 1));
  final period = DatePeriodAssembler(date, nextDay);
  // prints: "Number of days: 2"
  print('Number of days: ${period.inDays}');
  // print "Difference in days: 1"
  print('Difference in days: ${nextDay.difference(date).inDays}');
  // prints "Date is before nextDay: true"
  print('Date is before nextDay: ${date.isBefore(nextDay)}');
  // print "Date is after nextDay: false"
  print('Date is after nextDay: ${date.isAfter(nextDay)}');

  date = Date(2020, 05, 12);

  // print "toString(): 2020-05-12"
  print('toString(): $date');
  // print "Month (May): 5"
  print('Month (May): ${date.month}');

  /// before using toYMdString() we have to initialize intl
  try {
    print('Date in YMD format: ${date.toYMdString('en_us')}');
  } catch (e) {
    print(e);
  }

  // this must be done just once for the whole application for each locale used
  await initializeDateFormatting('en_us');
  // print "Date in YMd format: 5/12/2020"
  print('Date in YMD format: ${date.toYMdString('en_us')}');

  /// intl initialization (as above) must have been performed before
  /// calling toYMMMMdString()
  // print "Date in YMMMMd format: May 12, 2020"
  print('Date in YMMMMd format: ${date.toYMMMMdString('en_us')}');
}


lib/src/date_period_assembler.dart
import 'package:vy_date/src/date.dart';
import 'package:vy_date/src/date_period.dart';

// *******************************************************
// ******************** Assembler ************************
// *******************************************************
class DatePeriodAssembler {
  Date? _startDate;
  Duration? _duration;

  Date? get startDate => _startDate;

  set startDate(Date? value) {
    if (value == null) {
      if (_startDate == null) {
        return;
      }
      if (isValid()) {
        _duration = exclusiveEndDate?.difference(DatePeriod.startReference);
      }
      _startDate = null;
      return;
    }
    if (isValid()) {
      if (endDate! < value) {
        throw ArgumentError('Start Date must preceed End Date');
      }
      _duration = exclusiveEndDate?.difference(value);
    } else if (_duration != null) {
      Duration offsetDuration;
      offsetDuration = value.difference(DatePeriod.startReference);
      _duration = _duration! - offsetDuration;
    }
    _startDate = value;
  }

  Duration? get duration => _duration;

  /// setting a duration without a start date defaults from 1970-1-1
  set duration(Duration? value) {
    if (value == null) {
      _duration = null;
      return;
    }
    if (value.isNegative || value.inDays == 0) {
      throw ArgumentError(
          'Negative or zero duration not allowed in DatePeriod creation');
    }
    _duration = Duration(days: value.inDays);
  }

  Date? get exclusiveEndDate => _duration == null
      ? null
      : _startDate == null
          ? DatePeriod.startReference.add(_duration!)
          : _startDate!.add(_duration!);

  set exclusiveEndDate(Date? value) {
    if (value == null) {
      _duration = null;
      return;
    }
    if (isValid()) {
      if (value < _startDate!) {
        throw ArgumentError('Starting Date must preceed Ending Date');
      }
      _duration = value.difference(_startDate!);
    } else if (_startDate == null) {
      _duration = value.difference(DatePeriod.startReference);
    } else {
      if (_startDate!.isAfter(value)) {
        throw ArgumentError('Ending date must follow starting one');
      }
      _duration = value.difference(_startDate!);
    }
  }

  Date? get endDate => exclusiveEndDate?.subtract(DatePeriod.oneDay);
  set endDate(Date? value) => exclusiveEndDate = value?.add(DatePeriod.oneDay);

  int? get inDays => _duration?.inDays;

  DatePeriodAssembler([this._startDate, Date? inclusiveEndDate]) {
    endDate = inclusiveEndDate;
  }

  DatePeriodAssembler.byDuration(Date startDate, Duration duration) {
    _startDate = startDate;
    this.duration = duration;
  }

  DatePeriodAssembler.byPeriod(DatePeriod period)
      : _startDate = period.startDate,
        _duration = period.duration;

  @override
  String toString() => 'From $_startDate to $endDate ';

  String toYMMMMdString(String locale) =>
      '${_startDate?.toYMMMMdString(locale) ?? '\u00a0'} '
      '\u2796 ${endDate?.toYMMMMdString(locale) ?? '\u00a0'}';

  static DatePeriodAssembler parseYMMMMdString(
      String yMMMMdString, String locale) {
    final parts = yMMMMdString.split('\u2796');
    if (parts.length != 2) {
      throw ArgumentError(
          'Wrong input string. Cannot distinguish the two dates');
    }
    final start = Date.parseYMMMMdString(parts[0].trim(), locale);
    final end = Date.parseYMMMMdString(parts[1].trim(), locale);
    return DatePeriodAssembler(start, end);
  }

  void initFrom(DatePeriod period) {
    _startDate = period.startDate;
    _duration = period.duration;
  }

  DatePeriod generate() => isValid()
      ? DatePeriod.byDuration(_startDate!, _duration!)
      : throw ArgumentError('Invalid DatePeriod');

  static DatePeriodAssembler newFromMap(Map<String, dynamic> map) {
    return DatePeriodAssembler(Date.parse(map[DatePeriod.fieldStartDate]),
        Date.parse(map[DatePeriod.fieldEndDate]));
  }

  DatePeriodAssembler duplicate() => DatePeriodAssembler(_startDate, endDate);

  // ******** functional methods ***********

  bool isValid() => _startDate != null && _duration != null;

  @override
  bool operator ==(other) {
    return other is DatePeriodAssembler &&
        _startDate == other._startDate &&
        _duration == other._duration;
  }

  @override
  int get hashCode => Object.hash(_startDate.hashCode, _duration.hashCode);
}


lib/src/date_period.dart
import 'dart:convert';
import 'package:vy_date/vy_date.dart' show Date;

class DatePeriod implements Comparable<DatePeriod> {
  static const String fieldStartDate = 'startDate';
  static const String fieldDuration = 'duration';
  static const String fieldEndDate = 'endDate';
  static final Date startReference = Date(1970, 1, 1);
  static const Duration oneDay = Duration(days: 1);

  DatePeriod.byDuration(Date inclusiveStartDate, this.duration)
      : startDate = inclusiveStartDate {
    if (duration < oneDay) {
      throw ArgumentError('Start date must be lesser or equal to end date');
    }
  }

  factory DatePeriod(Date inclusiveStartDate, Date inclusiveEndDate) =>
      DatePeriod.byDuration(inclusiveStartDate,
          inclusiveEndDate.add(oneDay).difference(inclusiveStartDate));

  factory DatePeriod.parseYMMMMdString(String yMMMMdString, String locale) {
    final parts = yMMMMdString.split('\u2796');
    if (parts.length != 2) {
      throw ArgumentError(
          'Wrong input string. Cannot distinguish the two dates');
    }
    final start = Date.parseYMMMMdString(parts[0].trim(), locale);
    final end = Date.parseYMMMMdString(parts[1].trim(), locale);
    return DatePeriod(start, end);
  }

  factory DatePeriod.fromJson(Map<String, dynamic> map) => DatePeriod(
      Date.parse(map[fieldStartDate]), Date.parse(map[fieldEndDate]));

  factory DatePeriod.revive(String jsonString) =>
      DatePeriod.fromJson(json.decode(jsonString));

  final Date startDate;
  final Duration duration;

  Date get exclusiveEndDate => startDate.add(duration);

  Date get endDate => exclusiveEndDate.subtract(oneDay);

  int get inDays => duration.inDays;

  @override
  String toString() => 'From $startDate to $endDate ';

  @override
  bool operator ==(other) {
    return other is DatePeriod &&
        startDate == other.startDate &&
        duration == other.duration;
  }

  @override
  int get hashCode => Object.hash(startDate.hashCode, duration.hashCode);

  @override
  int compareTo(DatePeriod other) {
    final dateResult = startDate.compareTo(other.startDate);
    if (dateResult != 0) {
      return dateResult;
    }
    return duration.compareTo(other.duration);
  }

  bool operator <(DatePeriod other) => compareTo(other) < 0;

  bool operator <=(DatePeriod other) => compareTo(other) <= 0;

  bool operator >=(DatePeriod other) => compareTo(other) >= 0;

  bool operator >(DatePeriod other) => compareTo(other) > 0;

  Map<String, dynamic> toJson() =>
      <String, String>{fieldStartDate: '$startDate', fieldEndDate: '$endDate'};

  // ****************** Functional Methods ***************************

  String toYMMMMdString(String locale) => '${startDate.toYMMMMdString(locale)} '
      '\u2796 ${endDate.toYMMMMdString(locale)}';

  bool isInPeriod(Date date) => date >= startDate && date < exclusiveEndDate;

  String encode() => json.encode(this);

  DatePeriod duplicate() => DatePeriod.byDuration(startDate, duration);

  /// this method split the current period based on the month and day
  /// of the given Date. They are considered as end date (inclusive).
  ///
  /// Ex. a period 2015-07-09 => 2018-03-14, split by the date 2010-06-30
  /// returns the following periods:
  /// * 2015-07-09 => 2016-06-30
  /// * 2016-07-01 => 2017-06-30
  /// * 2017-07-01 => 2018-03-14
  List<DatePeriod> splitByEndDate(Date endDate) {
    final periods = <DatePeriod>[];
    var restPeriod = duplicate();
    var startPeriodDate = startDate.duplicate();
    var endPeriodDate = Date(startPeriodDate.year, endDate.month, endDate.day);

    if (startPeriodDate > endPeriodDate) {
      endPeriodDate = Date(endPeriodDate.year + 1, endDate.month, endDate.day);
    }
    do {
      if (restPeriod.endDate > endPeriodDate) {
        periods.add(DatePeriod(startPeriodDate, endPeriodDate));
        startPeriodDate = endPeriodDate.add(DatePeriod.oneDay);
        restPeriod = DatePeriod(startPeriodDate, restPeriod.endDate);
        endPeriodDate =
            Date(endPeriodDate.year + 1, endDate.month, endDate.day);
      } else {
        periods.add(DatePeriod(startPeriodDate, restPeriod.endDate));
        break;
      }
    } while (true);
    return periods;
  }
}
